### ISSUE TITLE
Use ECMAScript Modules instead of CommonJS

### ISSUE THREAD

**Issue by jkjustjoshing** [uswds/uswds#4515]
ECMAScript Modules (aka "ES6 Modules" or "ESM") is a module system that natively works in most browsers. Newer versions of Node.js also natively support it. The USWDS javascript code should switch from the current module system, CommonJS, to using ECMAScript Modules.

**What makes ESM better?**

 * Tools like Rollup and Webpack can tree-shake ESM dependencies, but can't tree-shake CommonJS dependencies. If we want to do the following:
    ```js
    import { accordion } from 'uswds/src/js/components'
    import domready from 'domready'
    domready(() => {
        accordion.on()
    })
    ```
    Webpack or Rollup knows that we only need the `accordion` component's javascript, and it can delete all other component javascript.
    Caveat: tree shaking can be finicky, and other changes may need to be made to get it fully working. Those changes could be fast-follows to the initial work of getting ESM working.
 * ESM can be transpiled down to CommonJS, and then run through browserify like we do today. But in a few years when browser support for ESM is sufficient, the transpilation can simply be turned off and we can switch to directly taking advantage of native browser module support.

**Additional context**

In some ways this is the javascript sister issue to #2854.

**Comment by mejiaj** [comment]
I've done some prototyping and have added notes in this [draft PR](https://github.com/uswds/uswds/pull/5083).

Fully adopting ESM would be a major breaking change, not only to users but for the core team as well. With the current state of ESM and node support, I would not recommend a full migration. Maybe in a few months we can look at this again, but in it's current state fully adopting ESM doesn't offer major benefits.

Looking at your accordion example, you can currently import accordion via:

```js
import accordion from "@uswds/uswds/js/usa-accordion";
```

**Comment by aduth** [comment]
I'd wonder if this necessarily has to be a breaking change. A transitionary step could be to write ESM by default, but continue to compile and distribute a CommonJS version. Backward compatibility could be maintained by a combination of (1) preserving existing file naming for CJS and using `.mjs` extension for ESM, and (2) using [conditional exports](https://nodejs.org/docs/latest/api/packages.html#conditional-exports)† to refer to ESM/CJS copies in `package.json` `exports`.

Personally, I bumped into this when trying to reference source code directly in the browser with a combination of `<script type="module">` and [import maps](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/importmap).

It'd be really nice (convenient, maybe-performant) to be able to use USWDS like...

```html
<script type="importmap">
  {
    "imports": {
      "@uswds/uswds": "/node_modules/@uswds/uswds/path/to/index.mjs",
      "@uswds/uswds/accordion": "/node_modules/@uswds/uswds/path/to/accordion/index.mjs"
    }
  }
</script>
<script type="module">
import { accordion } from '@uswds/uswds';
// or: import accordion from '@uswds/uswds/accordion';
// or: import { accordion } from '/node_modules/@uswds/uswds/path/to/index.mjs';
// etc.
accordion.on();
</script>
```

† Conditional exports does require Node.js v12.16.0 or newer, which is technically a [breaking change](https://github.com/uswds/uswds/blob/ab9e3fdca8e81531ab3204c8a2dce0a8bf99fede/package.json#L6), but [Node.js 10.x was end-of-life'd in 2021](https://github.com/nodejs/Release?tab=readme-ov-file#end-of-life-releases), so it seems reasonable?

**Comment by jc00ke** [comment]
We'd love to see official support for ES6 modules, especially because of how the all supported browsers (according to the 2% rule) [support `importmap`](https://caniuse.com/import-maps). This can allow a team to avoid an entire ecosystem of dependencies, which is nice way to reduce risk on a project. 4.0.0 could be a great time to introduce such a departure and we're not so far away.

**Comment by mejiaj** [comment]
We'll be focusing our module work in web components. Moving forward this is the approach we're going to try to use.

**Comment by jc00ke** [comment]
As long as I can use `importmap` I'm happy 😉

**Comment by annepetersen** [comment]
Yep: addressing this will be in https://github.com/uswds/uswds-elements -land (if it beats out everything else in priority order.)
